### ISSUE TITLE
[Bug]: Some error in `join`, seems relate to #927

### ISSUE THREAD

**Issue by luke396** [narwhals-dev/narwhals#929]
### Describe the bug

This is the result of query 13 from TPC-H.

```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/luke/codes/python/narwhals/tpch/execute/q13.py", line 9, in <module>
    print(q13.query(fn(customer), fn(orders)))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/narwhals/translate.py", line 717, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/tpch/queries/q13.py", line 12, in query
    customer_ds.join(orders, left_on="c_custkey", right_on="o_custkey", how="left")
  File "/home/luke/codes/python/narwhals/narwhals/dataframe.py", line 1925, in join
    return super().join(other, how=how, left_on=left_on, right_on=right_on, on=on)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/narwhals/dataframe.py", line 217, in join
    self._compliant_frame.join(
  File "/home/luke/codes/python/narwhals/narwhals/_pandas_like/dataframe.py", line 495, in join
    return self._from_native_frame(result_native.drop(columns=extra))
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/.venv/lib/python3.12/site-packages/pandas/core/frame.py", line 5581, in drop
    return super().drop(
           ^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/.venv/lib/python3.12/site-packages/pandas/core/generic.py", line 4788, in drop
    obj = obj._drop_axis(labels, axis, level=level, errors=errors)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/.venv/lib/python3.12/site-packages/pandas/core/generic.py", line 4830, in _drop_axis
    new_axis = axis.drop(labels, errors=errors)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luke/codes/python/narwhals/.venv/lib/python3.12/site-packages/pandas/core/indexes/base.py", line 7070, in drop
    raise KeyError(f"{labels[mask].tolist()} not found in axis")
KeyError: "['o'] not found in axis
```

### Steps or code to reproduce the bug

TPCH-Q13 can be reproduced, 

and similarly, Q16, Q17, and Q18 might also be affected, although not necessarily for the same reason.

### Expected results

Return as expected.

### Actual results

Error see above.

### Please run narwhals.show_version() and enter the output below.

```shell
System:
    python: 3.12.4 (main, Jul 25 2024, 22:42:01) [Clang 18.1.8 ]
executable: /home/luke/codes/python/narwhals/.venv/bin/python
   machine: Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.35

Python dependencies:
     narwhals: 1.6.3
       pandas: 2.2.2
       polars: 1.6.0
         cudf: 
        modin: 
      pyarrow: 17.0.0
        numpy: 2.1.0
```

**Comment by MarcoGorelli** [comment]
thanks @luke396 for the report

I can't reproduce this
```
(polars-api-compat-dev) marcogorelli@DESKTOP-U8OKFP3:~/polars-api-compat-dev/tpch$ python -m execute.q13
    c_count  custdist
2         0     50004
8        10      6668
12        9      6563
20       11      6004
9         8      5890
14       12      5600
6        13      5029
13       19      4805
1         7      4680
3        18      4531
7        20      4507
17       14      4473
19       15      4463
18       17      4445
5        16      4410
15       21      4168
10       22      3742
0         6      3273
11       23      3189
22       24      2700
23       25      2090
21        5      1957
26       26      1653
16       27      1177
4         4      1010
24       28       901
28       29       564
31        3       408
27       30       378
29       31       242
30       32       133
35        2       128
32       33        72
34       34        52
36       35        32
25       36        20
33        1        20
38       37         8
41       38         4
37       41         3
40       40         3
39       39         1
shape: (42, 2)
┌─────────┬──────────┐
│ c_count ┆ custdist │
│ ---     ┆ ---      │
│ u32     ┆ u32      │
╞═════════╪══════════╡
│ 0       ┆ 50004    │
│ 10      ┆ 6668     │
│ 9       ┆ 6563     │
│ 11      ┆ 6004     │
│ 8       ┆ 5890     │
│ …       ┆ …        │
│ 37      ┆ 8        │
│ 38      ┆ 4        │
│ 41      ┆ 3        │
│ 40      ┆ 3        │
│ 39      ┆ 1        │
└─────────┴──────────┘
```

Are you running this from your fork of Narwhals or from the latest install?

**Comment by FBruzzesi** [comment]
I can replicate it. Issue seems to be the removal of 

```py
if isinstance(left_on, str):
    left_on = [left_on]
if isinstance(right_on, str):
    right_on = [right_on]
```
which then result in "zipping" characters instead of column names in 

```py
for left_key, right_key in zip(left_on, right_on):
    ...
```

**Comment by MarcoGorelli** [comment]
interesting thanks - how are you running it? latest commit? latest pip version? i'm confused as to why it's not showing up for me

**Comment by FBruzzesi** [comment]
Latest commit via the editable install

**Comment by MarcoGorelli** [comment]
ok i just recreated my env and I can reproduce it, thanks both! not sure what had gone wrong before

ok looks like the issue is that our tests all had single-letter variable names and so this was never caught 🤦 sorry about that, i should've picked up on it during review, fix incoming

**Comment by MarcoGorelli** [comment]
fixed by #930 

thanks a tonne both for reporting and triaging it so quickly 🙏
